src/main/java/zone/rong/loliasm/common/modfixes/storagedrawers/mixins/ControllerDataMixin.java

```diff
@@ -0,0 +1,41 @@
+package zone.rong.loliasm.common.modfixes.storagedrawers.mixins;
+
+import com.jaquadro.minecraft.storagedrawers.block.tile.TileEntityController;
+import com.jaquadro.minecraft.storagedrawers.block.tile.tiledata.ControllerData;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+@Mixin(value = ControllerData.class, remap = false)
+public class ControllerDataMixin {
+    @Shadow
+    private BlockPos controllerCoord;
+
+    private TileEntityController controller;
+
+    @Inject(method = "getController", at = @At(value = "HEAD"), cancellable = true)
+    void getController(TileEntity host, CallbackInfoReturnable<TileEntityController> cir) {
+        if (controller != null) {
+            if (!controller.isInvalid()) {
+                cir.setReturnValue(controller);
+            }
+            controller = null;
+            host.markDirty();
+            cir.setReturnValue(null);
```

Comment:
Is this meant to `setReturnValue` twice? I think you have to manually exit the method even when calling `setReturnValue`

Reply:
from what i understood setReturnValue is enough
```
    public void setReturnValue(R returnValue) throws CancellationException {
        super.cancel();
        this.returnValue = returnValue;
    }
```

Reply:
`setReturnValue` just cancels the rest of the *original* method from running, but your mixin will keep running till you return from it.